### ISSUE TITLE
Typescript useDispatch 'expected 0-1 arguments, but got 2'

### ISSUE THREAD

**Issue by vjsingh** [CharlesStover/reactn#96]
I read through the similar recent issue from slugs99, but still can't get Typescript to work.

I'm trying to use the new useDispatch syntax which takes two arguments, like so:
```const dispatchFilters = useDispatch(setFilter, 'filters');```

I'm using the latest reactn (2.1.4). I went into types/provider.d.ts and it does appear that useDispatch is typed to take two arguments so I'm not sure what is wrong.

Perhaps something in my ```global.d.ts``` file is wrong? I'm using something similar to the default template.

**Comment by CharlesStover** [comment]
`types/provider.d.ts` contains the definition for the `Provider` Component returned by the `createProvider` helper function. It will be very similar to the helper functions themselves, but is not the definition being used by the `useDispatch` helper function here.

Can you share your `global.d.ts` file and your `setFilter` function? My best guess is your `interface State` in `global.d.ts` may not contain a `filters` key and/or `setFilter` is not returning the same type as `filters`. If this is a bug in ReactN, the definition in your `global.d.ts` file and `setFilter` function can help be debug it and get it patched.

**Comment by vjsingh** [comment]
```Global.d.ts```:
```
import 'reactn';

declare module 'reactn/default' {

  export interface Reducers {

    setFilter: (
      global: State,
      dispatch: Dispatch,
    ) => Pick<State, 'filters'>;
  }

  export interface State {
    count: number;
    filters: any;
  }
}
```

```
const INITIAL_FILTER_STATE = {
  user: { name: 'Professor Plum', isActive: false},
};

// Set the initial global state.
setGlobal({
  filters: INITIAL_FILTER_STATE,
  count: 0,
});

// Switched to using the single argument useDispatch, which is working.
// export function setFilter(filters, action) {
export function setFilter(state, dispatch, action) {
  const filters = state.filters;
  const isActive = !filters[action.filterType].isActive;

  switch (action.type) {
    case 'TOGGLE_FILTER' :
      return {
        filters: {
          ...filters,
          [action.filterType]: {
            ...action.filter,
            isActive,
          },
        },
      };
    default:
      return state;
  }
}

```

**Comment by CharlesStover** [comment]
There are two types of reducers that ReactN supports: global reducers and property reducers.

A global reducer mutates any property of the global state. In addition to an action, it receives the current global state and all other reducers as parameters. It returns a partial global state. Global reducers are added via `addReducer` so that they can be attached to the global state itself and used anywhere in your app (`useDispatch('reducerName')` and `this.dispatch.reducerName`).

```JavaScript
// Add myGlobalReducer to the global state.
function myGlobalReducer(global: State, dispatch: Dispatch, one: string, two: string) {
  return {
    propertyOne: one,
    propertyTwo: two,
  };
}
addReducer('myGlobalReducer', myGlobalReducer);

function MyComponent() {
  // Sets global.propertyOne to 'xxx' and global.propertyTwo to 'yyy':
  const dispatch = useDispatch('myGlobalReducer');
  dispatch('xxx', 'yyy');
  return null;
}
```

A property reducers mutates only one property on the global state. In addition to an action, it receives the current value of that property (not the other global state properties or the reducers). It returns a new value for that property. A property reducer is not attached to the global state itself. To use it, you import the property reducer directly and specify which property is being reduced. This is analogous to React's `useReducer` behavior.

```JavaScript
setGlobal({
  count: 0,
});

// Define myPropertyReducer
function myPropertyReducer(num, add, subtract) {
  return num + add - subtract;
}

function MyComponent() {
  // Add 5 to and subtract 3 from global.count
  const dispatch = useDispatch(myPropertyReducer, 'count');
  dispatch(5, 3);
  return null;
}
```

If you are using a property reducer, you do not need to add it to `interface Reducers` because it is not a part of the global state. You _import_ and use it as needed. This is in contrast to a global reducer, which you may add to the global state (`addReducer`) in one file (e.g. `App.tsx`), but TypeScript has no knowledge of its existence in a second file (e.g. `Button.tsx`) where it is used. The goal of `'reactn/default'` is to tell TypeScript the shape of interfaces that are defined outside of the files in which they are used. A property reducer is accessible from the file in which it is used, so it does not need to be specified in `'reactn/default'`.

Your `setFilter` function is accepting `dispatch` as a parameter and returning `{ filters: ... }`. As a property reducer, it will not receive `dispatch` and only needs to return the new value for `filters`.

I know that's a lot to take in, but the end goal is you have a choice of whichever implementation you prefer. There are no benefits or detriments to either. It's a personal preference.

You'll want one of these two methods, whichever you like more:
```JavaScript
// Global reducer
// Define and add this function only ONCE in your app.
// I recommend somewhere top-level, like src/index.tsx,
//   or wherever you call `setGlobal` to instantiate the state.
function setFilter(state, dispatch, action) {
  const filters = state.filters;
  const isActive = !filters[action.filterType].isActive;
  switch (action.type) {
    case 'TOGGLE_FILTER' :
      return {
        filters: {
          ...filters,
          [action.filterType]: {
            ...action.filter,
            isActive,
          },
        },
      };
    default:
      return null;
  }
}
addReducer('setFilter', setFilter);

// my-component.tsx
function MyComponent() {
  const setFilter = useDispatch('setFilter');
  setFilter({
    type: 'TOGGLE_FILTER',
    filterType: 'xyz',
  });
}
```

or:

```JavaScript
// Property reducer
// You likely want to only define this once and import it as needed across multiple files.
// e.g. import setFilter from 'src/reducers/set-filter';
// If it is only used in one file, you may define it in that one file. However you want to handle it.
function setFilter(filters, action) {
  const isActive = !filters[action.filterType].isActive;
  switch (action.type) {
    case 'TOGGLE_FILTER' :
      return {
        ...filters,
        [action.filterType]: {
          ...action.filter,
          isActive,
        },
      };
    default:
      return filters;
  }
}

function MyComponent() {
  const setFilter = useDispatch(setFilter, 'filters');
  setFilter({
    type: 'TOGGLE_FILTER',
    filterType: 'xyz',
  });
}
```

Let me know if you this adjustment does not solve your issue, or if you need further assistance. I would like for the concept of global and property reducers to be more intuitive to users, so your feedback is appreciated.

**Comment by vjsingh** [comment]
Hi Charles,

Thank you so much for the detailed response. That is very helpful and comprehensive, and helps me understand the different between global and property reducers.

I think I may have misled you in my last post. Creating my reducer as a property reducer results in the error in my original post, that useDispatch expects 0-1 arguments.

I successfully refactored my code to use a global reducer instead of a property reducer, and it is working as expected now. I wonder if it may be easier to chat directly somehow to see why I get the typescript error when attempting to use the property reducer solution, since I would like to go back to that solution if possible.

**Comment by vjsingh** [comment]
Also curious about this line:

```
Global reducers are added via addReducer so that they can be attached to the global state itself and used anywhere in your app (useDispatch('reducerName') and this.dispatch.reducerName).
```

I forgot to use addReducer, and yet my code appears to still be working fine?

**Comment by CharlesStover** [comment]
If you don't use `addReducer` but put your Reducer in your `global.d.ts` file, TypeScript won't complain at compile time, but it shouldn't do anything at runtime, because that function won't exist on the global state and isn't actually being returned by `useDispatch`.

There is a Discord channel dedicated to ReactN, linked at the top and bottom of the README. I can't guarantee that I am available to respond immediately, but I am responding in my free time. :)

**Comment by CharlesStover** [comment]
I'm going to close this since it was taken up in the Discord and now lacks context. If you still need help or have a clear Issue/reproduction, feel free to re-open or make a new issue with the context needed to resolve it. 😄
